Propagate static_if more completely.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35834 Propagate static_if more completely.**

This handles the cases we did not handle before in AND and OR statements:

    static_true || <unknown> -> static_true
    static_false && <unknown> -> static_false

Differential Revision: [D20801125](https://our.internmc.facebook.com/intern/diff/D20801125)

zdevito committed:
Propagate static_if more completely.

This handles the cases we did not handle before in AND and OR statements:

    static_true || <unknown> -> static_true
    static_false && <unknown> -> static_false

[ghstack-poisoned]

zdevito committed:
Update on "Propagate static_if more completely."


This handles the cases we did not handle before in AND and OR statements:

    static_true || <unknown> -> static_true
    static_false && <unknown> -> static_false

[ghstack-poisoned]

eellison reviewed (approved):
Nice. you have a lint error btw

zdevito committed:
Update on "Propagate static_if more completely."


This handles the cases we did not handle before in AND and OR statements:

    static_true || <unknown> -> static_true
    static_false && <unknown> -> static_false

Differential Revision: [D20801125](https://our.internmc.facebook.com/intern/diff/D20801125)

[ghstack-poisoned]
